README.md
# microViewer
<p align="center">
  <img width="285" height="267" src="https://github.com/isMTv/microViewer/blob/main/screens/mv_main.png">
</p>

Софт для получения удаленного доступа к ПК через собственный OpenSSH сервер. Написан на языке Python с использованием QT и сторонних программ UltraVNC, KiTTY.

### Возможности:
* Настроить под собственный OpenSSH сервер;
* Не требователен к пропускной способности сети Интеренет;
* Авторизация по ключам (OpenSSH);
* Может работать как сквозь NAT, так и внутри локальной сети;
* Доступ по ID/IP/DNS-имени ПК (выбирается в режимах);
* Адресная книга с возможностью поиска записей, может работать отдельно от основной программы;
* Создание профилей для настроек программы и записей Адресной книги;
* Автоматическое переподключение к OpenSSH-туннелю при потере доступа к Интернет;
* Работа в режиме сервера (для временного включения доступа);
* Работа в режиме службы (получить постоянный доступ, возможно подключиться к ПК на этапе LogonScreen);
* В режиме службы, когда ID/Пароль закреплены, пароль будет зашифрован средствами UltraVNC (Отобразится в окне MV на вкладке "хост" при повторном запуске);
* Передача файлов, чат, возможность выбрать экран (при наличии второго монитора);
* Автоматическое cохранение ID/Пароля от предыдущего запуска (сброс сохроненной пары осуществляется кнопкой обновить);
* Задать свои ID/Пароль (диапазон ID 19000-19999, не пересекается с основным 20000-60000);
* Одновременная работа Серверного/Клиентского подключения;
* Оставить серверную пару Логин/Ключ и запускать на ПК на которых будет работать только режим сервера;

### Сторонние компоненты:
* UltraVNC (https://github.com/ultravnc/ultravnc);
* KiTTY (https://github.com/cyd01/KiTTY);

### QT:
* x32_QT5: (Win7) - Windows 7sp1-11;
* x32_QT5: Windows 8.1-11;
* x64_QT6: Windows 10-11 (Support 4K Displays);


### VNC Viewer:
```
 - Файл настроек:
# App_microViewer\conf\confsets.ini
[Viewer]
quality_index=2
quality_presets=((16, 6), (16, 4), (16, 2), (16, 0))

 - Расшифровка значений:
# Encoding:
 - ZRLE = 16
 - ZYWRLE = 17 | u2 = 10
 - XZ = 18 | XZYW = 19

# Colors:
 - 0: Full Colors
 - 1: 256 Colors
 - 2: 64 Colors
 - 3: 8 Colors
 - 4: 8 Dark Colors
 - 5: 4 Grey Colors
 - 6: Black & White
 ```

### Настройка OpenSSH сервера;
```
 - Авторизация по ключам:
# useradd -m -U -s /bin/bash mv_server
# passwd mv_server
# useradd -m -U -s /bin/bash mv_client
# passwd mv_client
# su - mv_server
# su - mv_client
---
$ mkdir .ssh
$ cat > .ssh/authorized_keys << EOF - ENTER вставляем наш публичный ключ ENTER - EOF
$ chmod 700 .ssh
$ chmod 600 .ssh/authorized_keys
---
# usermod -s /bin/false mv_server
# usermod -s /bin/false mv_client

Первые 3 параметра позволят быстрее убить сессию клиента, на стороне которого произошел разрыв соединения.
Иначе, после переподключения к SSH, RD-Клиент не сможет подключиться к RD-Cерверу.
# tee /etc/ssh/sshd_config.d/microviewer.conf >/dev/null <<EOF
TCPKeepAlive yes
ClientAliveInterval 5
ClientAliveCountMax 2

Match User mv_server
    PasswordAuthentication no
    AllowTcpForwarding remote
    AllowStreamLocalForwarding no
    AllowAgentForwarding no
    GatewayPorts no
    X11Forwarding no
    PermitTunnel no
    ForceCommand echo "You've successfully authenticated, but server does not provide shell access."

Match User mv_client
    PasswordAuthentication no
    AllowTcpForwarding local
    AllowStreamLocalForwarding no
    AllowAgentForwarding no
    GatewayPorts no
    X11Forwarding no
    PermitTunnel no
    ForceCommand echo "You've successfully authenticated, but server does not provide shell access."
EOF
# systemctl restart sshd.service
```

### Сборка с помощью Nuitka:
https://nuitka.net/doc/user-manual.html
```
 - Необходимые пакеты:
# python.exe -m pip install --upgrade pip
# pip install pyqt6 / для x32 pyqt5
# pip install wheel
# pip install nuitka
# pip install zstandard ordered-set (orderedset)
# pip install pywin32

# nuitka -h, --help
(--standalone) или (--onefile)
 - k_tunnel.py + k_tunnel_helper.py
# nuitka bin/k_utils/k_tunnel.py --follow-imports --standalone --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico --remove-output --windows-company-name='open-networks.ru' --windows-file-description='microViewer - Kitty_Tunnel' --windows-file-version=1.3.7.0 --windows-product-name='Kitty_Tunnel'
# nuitka bin/k_utils/k_tunnel_helper.py --follow-imports --standalone --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico --remove-output --windows-company-name='open-networks.ru' --windows-file-description='microViewer - Kitty_Tunnel_Helper' --windows-file-version=1.3.7.0 --windows-product-name='Kitty_Tunnel_Helper'
 - main.py
# nuitka main.py --onefile --follow-imports --plugin-enable=pyqt6 --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico -o microViewer.exe --remove-output --windows-company-name='open-networks.ru' --windows-file-description='microViewer - Remote Access over SSH' --windows-file-version=1.3.7.0 --windows-product-name='microViewer'
# nuitka main.py --standalone --follow-imports --plugin-enable=pyqt6 --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico --remove-output --windows-company-name='open-networks.ru' --windows-file-description='microViewer - Remote Access over SSH' --windows-file-version=1.3.7.0 --windows-product-name='microViewer'

 - Компиляция. Если для x64, то в переменной среды "Path": пути для x64 должны быть над x32;
- \Python310\Scripts\
- \Python310\
# x32, поднимаем вверх..
- \Python310-32\Scripts\
- \Python310-32\

 - Windows 7 (x32):
python-3.8.10 (x32) для full updates
python-3.7.6 (x32) для no updates (Win7 SP1 [21.02.2011])
 - k_tunnel.py + k_tunnel_helper.py
# nuitka bin/k_utils/k_tunnel.py --follow-imports --standalone --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico --remove-output --windows-company-name="open-networks.ru" --windows-file-description="microViewer - Kitty_Tunnel" --windows-file-version=1.3.7.0 --windows-product-name="Kitty_Tunnel"
# nuitka bin/k_utils/k_tunnel_helper.py --follow-imports --standalone --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico --remove-output --windows-company-name="open-networks.ru" --windows-file-description="microViewer - Kitty_Tunnel_Helper" --windows-file-version=1.3.7.0 --windows-product-name="Kitty_Tunnel_Helper"
 - main.py
# nuitka main.py --follow-imports --plugin-enable=pyqt5 --standalone --windows-disable-console --windows-icon-from-ico=bin\ui\img\microViewer.ico --remove-output --windows-company-name="open-networks.ru" --windows-file-description="microViewer - Remote Access over SSH" --windows-file-version=1.3.7.0 --windows-product-name="microViewer"
```


changelog.txt
 - v1.0 (07.08.2022)
1. Brunhildr заменена на Gungnir;
2. Основные переменные настроек перенесены в модуль;
3. Добавлена возможность при доступе по IP, подключаться как по IP, так и по короткому имени и доменному имени;
4. Исправлено. Если в настройках изменить длинну пароля и нажать применить, то автоматически сгенерируется пароль - без ручного обновления;
5. Исправлено. В раширенных настройках, если закрепить или изменить ID/Пароль, то автоматически обновятся только измененные данные на вкладке хост. Если отключить закрепление, то данные также обновятся;
6. Адресная книга. Добавлена иконка при выключенном состоянии подключения;
7. Настроены модальные окна;

 - v1.1 (12.08.2022)
1. Местоположение основного родительского окна сохраняется при закрытии;
2. Дочерние окна наследуют местоположения родительских;

 - v1.2 (12.09.2022)
1. Добавлена кнопка очистки от введенных данных для ID, PW, Search;
2. В расширенных настройках в отключенном состоянии "Закр. ID/Пароль" теперь неактивны все элементы;
3. Возможность выделить сгенерированные ID/Пароль для копирования через CTRL+C;
4. Добавлено включение автозапуска для режима Сервер;
5. Исправлено местоположение и тип "окна-сообщения", когда при включенном режиме в настройки вносятся изменения кнопкой "Применить";
6. Исправлено сохранение настроек в конфигурационный файл. Параметр сохраняется в файл конфигурации только если он новый;
7. Мелкие правки интерфейса и логики программы;
8. Gungnir. Изменено качество видео на 2 (По умолчанию 3);

 - v1.3 (27.09.2022)
1. Добавлена поддерка профилей для адресной книги;
2. Добавлена поддержка профилей для настроек программы;
3. Добавлено отображение диалогового окна перед удалением элемента таблицы или профиля;
4. Добавлена проверка на отображение диалогового окна: если попытаться удалить
 - Несуществующую запись таблицы;
 - Последний объект профиля;
5. Добавлено удаление пробелов для добавляемых или изменяемых:
 - Записей таблицы, ID, PW;
 - Объектов профиля;
6. Добавлены проверки для предотвращения добавления пустых записей таблицы и объектов профилей;
7. Добавлена проверка предотвращающая добавление одинаковых объектов профилей;
8. Изменено. Если ввести данные в форму но не добавить их, то при следующем вводе форма очистится;
9. Для экономии места у главного окна удален StatusBar;
10. Исправлено. Если запустить программу без прав Администратора и включить автозапуск сервера: затем включить службу 
 - то будет запущен именно сервер, с ошибкой - необходимы права Администратора, а сам режим можно будет выключить только отключив автозапуск;
11. Добавлен вывод текущего локального IP-Адреса;
12. Мелкие правки и улучшения программы;
13. KiTTY обновлен до v0.76.0.11;

 - v1.3.1 (01.10.2022)
1. Изменено. Теперь, если используется режим "Сервера" и "Клиента", то при закрытии программы, автоматически выключится запущенный режим;
2. У главного окна удален неиспользуемый MenuBar;
3. Мелкие правки и улучшения программы;

 - v1.3.2 (02.10.2022)
1. Переработан запуск "Сервера" и "Клиентского подключения". Теперь можно запускать несколько экземпляров microViewer с данными режимами;
2. Режим Службы. Исправлен запуск туннеля через планировщик. Когда отключен сетевой интерфейс:
 - Отключена опция. Запускать только при подключении к следующей сети: Любое подключение;

 - v1.3.3 (10.10.2022)
1. Добавлена возможность из "Адресной книги" двойным нажатием по строке подключаться/отключаться от удаленного клиента;
2. Мелкие правки и улучшения программы;

 - v1.3.4 (25.10.2022)
1. Мелкие правки и улучшения программы;

 - v1.3.5 (19.01.2023)
1. Gungnir заменен на UltraVNC v1.4.0.6;
2. Обновлены компоненты - Nuitka 1.3.8, KiTTY 0.76.1.2;
3. Дабавлен вывод версии программы в заголовок окна. Переименована кнопка включения/отключения режимов работы;

 - v1.3.6 (03.04.2023)
1. Обновлены компоненты - UltraVNC 1.4.0.9, KiTTY 0.76.1.5;
2. Исправлена ситуация, когда при переключении между профилями, при выключенной фиксации в формы ID и Пароль не подгружались значения;
3. Исправлена ситуация, когда при включении "режима сервера" задавался пароль "Encrypt". "Encrypt" подгружает предыдущий зашифрованный пароль;
4. Добавлена служба Helper, которая по таймеру будет проверять активность установленного подключения к туннелю. ( т.к. в очень редких случаях, при плохом интернете, туннель может автомаически не переподключаться);
5. Из расширенных настроек удален раздел "Электропитание" т.к. в VNC Server уже присутствует настройка предотвращающаяя включение ScreenSaver;
6. Добавлена возможность выбрать и настроить пресет качества отображаемой картинки в VNC Viewer;
7. Мелкие правки и улучшения программы;

 - v1.3.7 (09.06.2023)
1. Обновлены компоненты - UltraVNC 1.4.2.0 , KiTTY 0.76.1.12;
2. Добавлено автоматическое определение версий для "UltraVNC" и "KiTTY" на вклакдке "Info" -> "Сборка";
3. Убран вывод иконки в трей для KiTTY при использовании режима "Сервера" и "Клиента";
4. Добавлено удаление пробелов из введенного пароля при использовании режима "Клиента" и "Фиксации" в расширенных настройках;
5. Ultravnc-Server включен по дефолту RDPmode (возможность подключиться к RDP-сессии);
6. Ultravnc-Client задан по дефолту наиболее стабильный пресет качества с использованием сжатия ZRLE;
7. Мелкие правки и улучшения программы;


cversion.py
from os import chdir
from pathlib import Path
from platform import python_version, architecture
from win32api import GetFileVersionInfo, HIWORD, LOWORD

mv_version = "1.3.7"
qt_version = "6.3.1" # 5.15.7 | 6.3.1
nuitka_version = "1.3.8"

dir_path = Path.cwd()
chdir(dir_path)
path_kitty = Path(f"bin/kitty.exe")
path_uvnc = Path(f"bin/uvnc/winvnc.exe")
arc = architecture()


def set_title_app_version(window: any) -> None:
    window.setWindowTitle(f"microViewer v{mv_version}")


def get_app_version(app_path: str) -> str:
    try:
        file_info = GetFileVersionInfo(app_path, "\\")
        ms_file_ver = file_info['FileVersionMS']
        ls_file_ver = file_info['FileVersionLS']
        # major.minor.build.revision;
        ver_mmbr = (str(HIWORD(ms_file_ver)), str(LOWORD(ms_file_ver)), str(HIWORD(ls_file_ver)), str(LOWORD(ls_file_ver)))

        ver = ".".join(ver_mmbr)
    except Exception:
        ver = "N/A"
    return ver


def set_components_version(form: any) -> None:
    form.label_mv_v.setText(f"v{mv_version} ({arc[0]})")
    form.label_python_v.setText(python_version())
    form.label_qt_v.setText(f"{qt_version}")
    form.label_nuitka_v.setText(f"{nuitka_version}")
    form.label_kitty_v.setText(f"{get_app_version(str(path_kitty))}")
    form.label_ultravnc_v.setText(f"{get_app_version(str(path_uvnc))}")


set_vars.py
import re
from secrets import choice
from random import randint
from string import ascii_letters, digits
from os import chdir
from pathlib import Path
from configs import createconfig_confsets, createconfig_mv, createconfig_rd, createconfig_viewer, get_config

# Data Path VARS;
dir_path = Path.cwd()
chdir(dir_path)

# Файл настроек ConfSets;
cs_conf_path = Path("conf/confsets.ini")
if not Path(cs_conf_path).exists():
    cs_conf_path.parent.mkdir(parents=False, exist_ok=True)
    createconfig_confsets(cs_conf_path)


class ConfSets:
    index: int = int(get_config(cs_conf_path, "ConfSets", "index"))
    profiles: list = eval(get_config(cs_conf_path, "ConfSets", "profiles"))
    all: tuple = ("index", "profiles")


class Viewer:
    quality_index: int = int(get_config(cs_conf_path, "Viewer", "quality_index"))
    quality_presets: tuple = eval(get_config(cs_conf_path, "Viewer", "quality_presets"))
    all: tuple = ("quality_index", "quality_presets")


# Файл настроек MV;
mv_conf_path = Path(f"conf/{ConfSets.profiles[ConfSets.index]}/mv.ini")
if not Path(mv_conf_path).exists():
    mv_conf_path.parent.mkdir(parents=False, exist_ok=True)
    createconfig_mv(mv_conf_path)

# Файл настроек RD;
rd_conf_path = Path("bin/uvnc/UltraVNC.ini")
if not Path(rd_conf_path).exists():
    rd_conf_path.parent.mkdir(parents=True, exist_ok=True)
    createconfig_rd(rd_conf_path)

# Файл настроек Viewer;
viewer_conf_path = Path("bin/uvnc/viewer.vnc")
if not Path(viewer_conf_path).exists():
    viewer_conf_path.parent.mkdir(parents=True, exist_ok=True)
    createconfig_viewer(viewer_conf_path)

# App's Path;
rd_app_server = Path(dir_path, "bin/uvnc/winvnc.exe")
rd_app_client = Path(dir_path, "bin/uvnc/vncviewer.exe")
k_app = "bin/kitty.exe"
k_tunnel = Path(dir_path, "bin/k_utils/k_tunnel.exe")
k_tunnel_helper = Path(dir_path, "bin/k_utils/k_tunnel_helper.exe")
k_task_tunnel = Path(dir_path, "bin/tasks/k_tunnel.xml")
k_task_tunnel_helper = Path(dir_path, "bin/tasks/k_tunnel_helper.xml")


class MV:
    password_length: int = int(get_config(mv_conf_path, "MV", "password_length"))
    connection_type: str = get_config(mv_conf_path, "MV", "connection_type")
    mode_server_autorun: str = get_config(mv_conf_path, "MV", "mode_server_autorun")
    all: tuple = ("password_length", "connection_type", "mode_server_autorun")


class ABOOK:
    index: int = int(get_config(mv_conf_path, "ABOOK", "index"))
    profiles: list = eval(get_config(mv_conf_path, "ABOOK", "profiles"))
    all: tuple = ("index", "profiles")


class Tunnel:
    host: str = get_config(mv_conf_path, "Tunnel", "host")
    port: int = int(get_config(mv_conf_path, "Tunnel", "port"))
    user_server: str = get_config(mv_conf_path, "Tunnel", "user_server")
    key_server: str = get_config(mv_conf_path, "Tunnel", "key_server")
    user_client: str = get_config(mv_conf_path, "Tunnel", "user_client")
    key_client: str = get_config(mv_conf_path, "Tunnel", "key_client")
    all: tuple = ("host", "port", "user_server", "key_server", "user_client", "key_client")


class RemoteDesktop:
    portnumber: int = get_config(rd_conf_path, "admin", "portnumber")
    all: tuple = ("portnumber",)


class Server:
    id: str = get_config(mv_conf_path, "Server", "id")
    password: str = get_config(mv_conf_path, "Server", "password")
    all: tuple = ("id", "password")


class Service:
    enabled: str = get_config(mv_conf_path, "Service", "enabled")
    check_interval: int = int(get_config(mv_conf_path, "Service", "check_interval"))
    helper: str = (get_config(mv_conf_path, "Service", "helper")).lower()
    check_interval_helper: int = int(get_config(mv_conf_path, "Service", "check_interval_helper"))
    fix_id_pw: str = (get_config(mv_conf_path, "Service", "fix_id_pw")).lower()
    id: str = get_config(mv_conf_path, "Service", "id")
    password: str = get_config(mv_conf_path, "Service", "password")
    all: tuple = ("enabled", "check_interval", "helper", "check_interval_helper", "fix_id_pw", "id", "password")


class Modes:
    service: str = get_config(mv_conf_path, "Modes", "service")
    all: tuple = ("service",)


class Pids:
    k_tunnel: str = get_config(mv_conf_path, "Pids", "k_tunnel")
    all: tuple = ("k_tunnel",)


class ParentWindow:
    position_x: int = get_config(mv_conf_path, "ParentWindow", "position_x")
    position_y: int = get_config(mv_conf_path, "ParentWindow", "position_y")
    all: tuple = ("position_x", "position_y")


# Если параметр обновился в файле конфига, то обновляем его внутри класса;
def refresh_classes_vars(conf_path: any, data_class: any) -> None:
    for index in eval(data_class).all:
        data = get_config(conf_path, data_class, index)
        command = f'{data_class}.{index} = "{data}"'
        exec(command)


# Чтобы не ломался eval(tuple and list);
def refresh_profiles(conf_path: any, data_class: any) -> None:
    for index in eval(data_class).all:
        data = get_config(conf_path, data_class, index)
        command = f'{data_class}.{index} = {data}'
        exec(command)


def refresh_uvnc(conf_path: any, data_class: any) -> None:
    for index in eval(data_class).all:
        data = get_config(conf_path, f"admin", index)
        command = f'{data_class}.{index} = {data}'
        exec(command)


class Rand_Id_Pw:
    # Генерация рандомного ID в диапазоне;
    @staticmethod
    def id() -> any:
        global mv_conf_path
        rand_id = "None"
        if MV.connection_type == "id":
            # Перед обновлением, узнаем с какого конфига брать значения;
            mv_conf_path = Path(f"conf/{ConfSets.profiles[ConfSets.index]}/mv.ini")
            refresh_classes_vars(mv_conf_path, "Service")
            if Service.enabled == "false" and Service.fix_id_pw == "false":
                # Подгружаем из конфига предыдущее значение;
                if not Server.id:
                    rand_id = str(randint(20000, 60000))
                else:
                    rand_id = Server.id
            else:
                rand_id = Service.id
        elif MV.connection_type == "ip":
            rand_id = "N/A"
        return rand_id

    # Генерация рандомного пароля;
    @staticmethod
    def pw() -> str:
        global mv_conf_path
        password_length = int(MV.password_length)
        # Перед обновлением, узнаем с какого конфига брать значения;
        mv_conf_path = Path(f"conf/{ConfSets.profiles[ConfSets.index]}/mv.ini")
        refresh_classes_vars(mv_conf_path, "Service")
        if Service.enabled == "false" and Service.fix_id_pw == "false":
            # Подгружаем из конфига предыдущее значение;
            if not Server.password:
                # Убираем похожие символы;
                let_dig = re.sub("[10IoOl]", "", ascii_letters + digits)
                while True:
                    rand_pw = ''.join(choice(let_dig) for i in range(password_length))
                    if (any(c.islower() for c in rand_pw)
                            and any(c.isupper() for c in rand_pw)
                            and sum(c.isdigit() for c in rand_pw) >= 3):
                        break
            else:
                rand_pw = Server.password
        else:
            rand_pw = Service.password
        return rand_pw


configs.py
import configparser


# Создание конфига для ConfSets;
def createconfig_confsets(conf_path: any) -> None:
    config = configparser.ConfigParser()
    config.add_section("ConfSets")
    config.set("ConfSets", "index", "0")
    config.set("ConfSets", "profiles", "['main']")
    config.add_section("Viewer")
    config.set("Viewer", "quality_index", "2")
    config.set("Viewer", "quality_presets", "((16, 6), (16, 4), (16, 2), (16, 0))")
    with open(conf_path, "w") as config_file:
        config.write(config_file, space_around_delimiters=False)


# Создание конфига для MV;
def createconfig_mv(conf_path: any) -> None:
    config = configparser.ConfigParser()
    config.add_section("MV")
    config.set("MV", "password_length", "7")
    config.set("MV", "connection_type", "id")
    config.set("MV", "mode_server_autorun", "false")
    config.add_section("ABOOK")
    config.set("ABOOK", "index", "0")
    config.set("ABOOK", "profiles", "['abook']")
    config.add_section("Tunnel")
    config.set("Tunnel", "host", "")
    config.set("Tunnel", "port", "22")
    config.set("Tunnel", "user_server", "mv_server")
    config.set("Tunnel", "key_server", "bin/keys/mv_server.ppk")
    config.set("Tunnel", "user_client", "mv_client")
    config.set("Tunnel", "key_client", "bin/keys/mv_client.ppk")
    config.add_section("Server")
    config.set("Server", "id", "")
    config.set("Server", "password", "")
    config.add_section("Service")
    config.set("Service", "enabled", "false")
    config.set("Service", "check_interval", "5")
    config.set("Service", "helper", "false")
    config.set("Service", "check_interval_helper", "180")
    config.set("Service", "fix_id_pw", "false")
    config.set("Service", "id", "19000")
    config.set("Service", "password", "")
    config.add_section("Modes")
    config.set("Modes", "service", "false")
    config.add_section("Pids")
    config.set("Pids", "k_tunnel", "none")
    config.add_section("ParentWindow")
    config.set("ParentWindow", "position_x", "none")
    config.set("ParentWindow", "position_y", "none")
    with open(conf_path, "w") as config_file:
        config.write(config_file, space_around_delimiters=False)


# Создание конфига для RD;
def createconfig_rd(conf_path: any) -> None:
    config = configparser.ConfigParser()
    config.add_section("admin")
    config.set("admin", "PortNumber", "5900")
    config.set("admin", "HTTPConnect", "0")
    config.set("admin", "AutoPortSelect", "0")
    config.set("admin", "MaxViewers", "5")
    config.set("admin", "AllowShutdown", "0")
    config.set("admin", "AllowEditClients", "0")
    config.set("admin", "FileTransferTimeout", "60")
    config.set("admin", "noscreensaver", "1")
    config.set("admin", "rdpmode", "1")
    config.add_section("UltraVNC")
    config.set("UltraVNC", "passwd", "")
    config.set("UltraVNC", "passwd2", "")
    with open(conf_path, "w") as config_file:
        config.write(config_file, space_around_delimiters=False)


# Создание конфига для Viewer;
def createconfig_viewer(conf_path: any) -> None:
    config = configparser.ConfigParser()
    config.add_section("options")
    config.set("options", "preferred_encoding", "16")
    config.set("options", "8bit", "2")
    config.set("options", "directx", "1")
    config.set("options", "JapKeyboard", "1")
    config.set("options", "Reconnect", "12")
    config.set("options", "AutoReconnect", "10")
    config.set("options", "filetransfertimeout", "60")
    with open(conf_path, "w") as config_file:
        config.write(config_file, space_around_delimiters=False)


# Редактирование конфига;
def edit_config(conf_path: any, section: str, settings: str, value: any) -> None:
    config = configparser.ConfigParser()
    config.read(conf_path)
    config.set(section, settings, value)
    with open(conf_path, "w") as config_file:
        config.write(config_file, space_around_delimiters=False)


# Получение значений;
def get_config(conf_path: any, section: str, settings: str) -> any:
    config = configparser.ConfigParser()
    config.read(conf_path)
    return config.get(section, settings)


# Удаление значений;
def remove_config(conf_path: any, section: str, settings: str) -> None:
    config = configparser.ConfigParser()
    config.read(conf_path)
    config.remove_option(section, settings)
    with open(conf_path, "w") as config_file:
        config.write(config_file, space_around_delimiters=False)
